src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -696,89 +617,159 @@ HeapTupleOfForeignConstraintIncludesColumn(HeapTuple heapTuple, Oid relationId,
- * search, therefore call this function with caution.
+ * TableReferencing function checks whether given table is referencing to another
+ * table via foreign key constraints. If it is referencing, this function returns
+ * true.
  */
 bool
-TableReferencing(Oid relationId)
+TableReferencing(Oid relationOid)
 {
-	ScanKeyData scanKey[1];
-	int scanKeyCount = 1;
-	Oid scanIndexId = InvalidOid;
-	bool useIndex = false;
+	int flags = EXTRACT_REFERENCING_CONSTRAINTS;
+	List *tableReferencingForeignConstraintOids = GetTableForeignConstraintOidsInternal(
+		relationOid, flags);
 
-	Relation pgConstraint = heap_open(ConstraintRelationId, AccessShareLock);
+	return list_length(tableReferencingForeignConstraintOids) > 0;
+}
 
-	ScanKeyInit(&scanKey[0], Anum_pg_constraint_conrelid, BTEqualStrategyNumber, F_OIDEQ,
-				relationId);
-	SysScanDesc scanDescriptor = systable_beginscan(pgConstraint, scanIndexId, useIndex,
-													NULL,
-													scanKeyCount, scanKey);
 
-	HeapTuple heapTuple = systable_getnext(scanDescriptor);
-	while (HeapTupleIsValid(heapTuple))
+/*
+ * ConstraintIsAForeignKey returns true if the given constraint name
+ * is a foreign key defined on the relation.
+ */
+bool
+ConstraintIsAForeignKey(char *inputConstaintName, Oid relationOid)
+{
+	int flags = EXTRACT_REFERENCING_CONSTRAINTS;
+	List *foreignKeyOids = GetTableForeignConstraintOidsInternal(relationOid, flags);
+
+	Oid foreignKeyOid = FindForeignKeyOidWithName(foreignKeyOids, inputConstaintName);
+
+	return OidIsValid(foreignKeyOid);
+}
+
+
+/*
+ * FindForeignKeyOidWithName searches the foreign key constraint with
+ * inputConstraintName in the given list of foreign key constraint OIDs.
+ * Returns the OID of the matching constraint. If there no matching constraint
+ * in the given list, then returns InvalidOid.
+ */
+static Oid
+FindForeignKeyOidWithName(List *foreignKeyOids, const char *inputConstraintName)
+{
+	Oid foreignKeyOid = InvalidOid;
+	foreach_oid(foreignKeyOid, foreignKeyOids)
 	{
-		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
+		char *constraintName = get_constraint_name(foreignKeyOid);
 
-		if (constraintForm->contype == CONSTRAINT_FOREIGN)
+		if (strncmp(constraintName, inputConstraintName, NAMEDATALEN) == 0)
 		{
-			systable_endscan(scanDescriptor);
-			heap_close(pgConstraint, NoLock);
-
-			return true;
+			return foreignKeyOid;
 		}
-
-		heapTuple = systable_getnext(scanDescriptor);
 	}
 
-	systable_endscan(scanDescriptor);
-	heap_close(pgConstraint, NoLock);
-
-	return false;
+	return InvalidOid;
 }
 
 
 /*
- * ConstraintIsAForeignKey returns true if the given constraint name
- * is a foreign key to defined on the relation.
+ * GetTableForeignConstraintOidsInternal takes in a relationOid, and returns
+ * the list of OIDs for foreign constraints that the relation with relationOid
+ * is involved according to "flags" argument. In that sense,
```

Comment:
minor: Could be nice to mention here that the current limitation is only because we have not implemented handling both. There is no theoretical blocker